src/db.rs

```diff
@@ -58,47 +70,56 @@ pub struct SingleThreaded {
     cfs: BTreeMap<String, ColumnFamily>,
 }
 
-/// Actual marker type for the internal marker trait `ThreadMode`, which holds
+/// Actual marker type for the marker trait `ThreadMode`, which holds
 /// a collection of column families wrapped in a RwLock to be mutated
 /// concurrently. The other mode is [`SingleThreaded`].
 ///
 /// See [`DB`] for more details, including performance implications for each mode
 pub struct MultiThreaded {
-    cfs: RwLock<BTreeMap<String, ColumnFamily>>,
+    cfs: RwLock<BTreeMap<String, Arc<UnboundColumnFamily>>>,
 }
 
 impl ThreadMode for SingleThreaded {
-    fn new(cfs: BTreeMap<String, ColumnFamily>) -> Self {
-        Self { cfs }
+    fn new_cf_map_internal(
+        cfs: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>,
+    ) -> Self {
+        Self {
+            cfs: cfs
+                .into_iter()
+                .map(|(n, c)| (n, ColumnFamily { inner: c }))
+                .collect(),
+        }
     }
 
-    fn cf_drop_all(&mut self) {
-        for cf in self.cfs.values() {
-            unsafe {
-                ffi::rocksdb_column_family_handle_destroy(cf.inner);
-            }
-        }
+    fn drop_all_cfs_internal(&mut self) {
+        // Cause all ColumnFamily objects to be Drop::drop()-ed.
+        self.cfs.clear();
     }
 }
 
 impl ThreadMode for MultiThreaded {
-    fn new(cfs: BTreeMap<String, ColumnFamily>) -> Self {
+    fn new_cf_map_internal(
+        cfs: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>,
+    ) -> Self {
         Self {
-            cfs: RwLock::new(cfs),
+            cfs: RwLock::new(
+                cfs.into_iter()
+                    .map(|(n, c)| (n, Arc::new(UnboundColumnFamily { inner: c })))
+                    .collect(),
+            ),
         }
     }
 
-    fn cf_drop_all(&mut self) {
-        for cf in self.cfs.read().unwrap().values() {
-            unsafe {
-                ffi::rocksdb_column_family_handle_destroy(cf.inner);
-            }
-        }
+    fn drop_all_cfs_internal(&mut self) {
+        // Cause all UnboundColumnFamily objects to be Drop::drop()-ed.
+        self.cfs.write().unwrap().clear();
```

Comment:
`unwrap()` -> `expect("Couldn't acquire a write lock")` ?

Reply:
thanks for carefully-thought suggestion.

However, according to https://doc.rust-lang.org/std/sync/struct.RwLock.html#errors-2, RwLock::write() return Err if and only if it's poisoned.

> This function will return an error if the RwLock is poisoned. An RwLock is poisoned whenever a writer panics while holding an exclusive lock. An error will be returned when the lock is acquired.

Also, RwLock::write() is blocking indefinitely. so your suggested message could be confusing and just better off propagating the RwLock Err as-is with `.unwrap()`?:

> Locks this rwlock with exclusive write access, blocking the current thread until it can be acquired.
> This function will not return while other writers or other readers currently have access to the lock.

Reply:
oops the above comment assumes you read my other comment here: https://github.com/rust-rocksdb/rust-rocksdb/pull/509/files#r674138005